### PR TITLE
Task d   cart creation

### PR DESCRIPTION
**Pull request by adriancorcoran**
## Aim
Adds a cart to the store

## How?
1. Session holds the cart id
1. New display page for the cart, using partials
1. Update tests

## Notes
Nope

### CODE REVIEW

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      format.json { head :no_content }
+    end
+  end
+
+  # get total number of items in the cart or 0
+  def total_num_items
+    if @cart.line_items
+      return @cart.line_items.count
+    end
+    0
```
Comment (jesseschroder) [inline app/controllers/carts_controller.rb]: With rails the `line_items` association is created whether or not there are any line items. So this should be able to be simplified

```suggestion
    @cart.line_items.count
```

Outside Rails, if you were creating an instance of a Ruby class that had an attribute with other objects in it, then yes you would need to check if it exists first. Unless that attribute is initialized as an empty array by default of course, which is basically what rails is doing with a `ActiveRecord::Associations::CollectionProxy` object.

File: app/controllers/carts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  end
+  
+  # get total cost of items in the cart or 0
+  def total_cost_items
```
Comment (jesseschroder) [inline app/controllers/carts_controller.rb:72]: Same idea here.

File: app/controllers/products_controller.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ class ProductsController < ApplicationController
   # GET /products
   # GET /products.json
   def index
-    @products = Product.all
+    @products = Product.all.order(:title)
```
Comment (jesseschroder) [inline app/controllers/products_controller.rb:7]:
```suggestion
    @products = Product.order(:title)
```
Simplified but same query.

File: app/controllers/concerns/current_cart.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  
+  private
+
+    def set_cart
+      @cart = Cart.find(session[:cart_id])
+    rescue ActiveRecord::RecordNotFound
+      @cart = Cart.create 
+      session[:cart_id] = @cart.id
+    end 
+end
```
Comment (jesseschroder) [inline app/controllers/concerns/current_cart.rb]: Missing EOF blank line

File: app/views/carts/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<% content_for :main_heading_action_button do %>
+  <%= render "shared/button", text: "Back", link: products_path, style_class: 'Polaris-Button--primary' %>
+<% end %>
+
+<%= render 'shared/page_title', title: "Your Cart" %>
+<%= render 'table', title: "Your Cart" %>
```
Comment (jesseschroder) [inline app/views/carts/show.html.erb]: Missing EOF line.

**Review by jesseschroder — COMMENTED** [review]
Good shit dude! Nice use of Polaris. I left a few comments regarding simplification of code, and a couple missing eof but looks awesome!

File: app/controllers/carts_controller.rb (hunk 1 above)
Comment (adriancorcoran) [reply app/controllers/carts_controller.rb]: Cool Jesse thank you for this. I added it in as I ran into an issue when testing and this seemed to solve it, however I think the issue in the tests eventually turned out to be something else, and I didn't remove this. I'll take it out and test :)

**Comment by adriancorcoran** [comment]
Awesome, thanks Jesse - TIL :)

**Pull request merged**